### PR TITLE
Access sharingLinkToRedeem from IOdspResolvedUrl.shareLinkInfo

### LINKED ISSUE #8250: redeemSharingLink throws when it shouldn't
We're seeing some cases in telemetry where odsp-driver's `fetchSnapshotWithRedeem` will throw during fallback because it thinks the share link isn't present in the resolved url, even though it makes the same check before doing fallback.  This appears to be because the share link was moved in the resolved url in #7554 - the deprecated property is used to check if fallback should happen (in `isRedeemSharingLinkError`), whereas the new property is checked within the fallback itself.  If one is present, we would expect the other to be present too.  So the suspicion is that as part of that deprecation, code paths were not all updated correctly and we end up in this state.

### LINKED ISSUE #8268: Access sharingLinkToRedeem from IOdspResolvedUrl.shareLinkInfo

### PR DESCRIPTION
**Pull request by sonalivdeshpande**
This PR aims to replace one of the deprecated property `sharingLinkToRedeem` from `IOdspResolvedUrl` interface. It can be accessed from `IOdspResolvedUrl.shareLinkInfo`. All the instances where this property was accessed from `IOdspResolvedUrl` interface is replaced with `IOdspResolvedUrl.shareLinkInfo.sharingLinkToRedeem`. 

Fixes: https://github.com/microsoft/FluidFramework/issues/8268 and https://github.com/microsoft/FluidFramework/issues/8250

### CODE REVIEW

File: packages/drivers/odsp-driver/src/fetchSnapshot.ts [hunk 1]
```diff
@@ -104,7 +104,7 @@ export async function fetchSnapshotWithRedeem(
             }, error);
             await redeemSharingLink(odspResolvedUrl, storageTokenFetcher, logger);
             const odspResolvedUrlWithoutShareLink: IOdspResolvedUrl =
-                { ...odspResolvedUrl, sharingLinkToRedeem: undefined };
```
Comment (jatgarg) [inline packages/drivers/odsp-driver/src/fetchSnapshot.ts:107]: `            const odspResolvedUrlWithoutShareLink: IOdspResolvedUrl =
                { ...odspResolvedUrl, shareLinkInfo: {...odspResolvedUrlWithoutShareLink.shareLinkInfo,
                    sharingLinkToRedeem: undefined,
                };
            }`
i think it should be like this and then you can remove the below if.

Comment (jatgarg) [reply packages/drivers/odsp-driver/src/fetchSnapshot.ts:107]: @pragya91 to confirm.

File: packages/drivers/odsp-driver/src/test/odspDriverUrlResolverForShareLink.spec.ts [hunk 2]
```diff
@@ -182,7 +182,7 @@ describe("Tests for OdspDriverUrlResolverForShareLink resolver", () => {
             return urlResolverWithShareLinkFetcher.resolve(
                 { url: url.toString(), headers: { [SharingLinkHeader.isSharingLinkToRedeem]: true } });
         });
-        assert(resolvedUrl.sharingLinkToRedeem !== undefined, "Sharing link should be set in resolved url");
+        assert(resolvedUrl.shareLinkInfo?.sharingLinkToRedeem !== undefined, "Sharing link should be set in resolved url");
         assert(resolvedUrl.shareLinkInfo?.sharingLinkToRedeem !== undefined, "Sharing link should be set in resolved url");
     });
```
Comment (pragya91) [inline packages/drivers/odsp-driver/src/test/odspDriverUrlResolverForShareLink.spec.ts:187]: Duplicate assert.

**Comment by agarwal-navin** [comment]
@sdeshpande3 This looks related to your PR - https://github.com/microsoft/FluidFramework/issues/8250.
Will your PR fix it? If not, can you take a look at that one too?

**Comment by agarwal-navin** [comment]
nit: `fron` -> `from` in PR title

**Comment by sdeshpande3** [comment]
> @sdeshpande3 This looks related to your PR - #8250. Will your PR fix it? If not, can you take a look at that one too?

Yes, #8250 is also resolved.

**Review by ChumpChief — APPROVED** [review]

**Pull request merged**